
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/dataplane/standalone/proto;sai";

enum IpmcEntryAttr {
  IPMC_ENTRY_ATTR_UNSPECIFIED = 0;
  IPMC_ENTRY_ATTR_PACKET_ACTION = 1;
  IPMC_ENTRY_ATTR_OUTPUT_GROUP_ID = 2;
  IPMC_ENTRY_ATTR_RPF_GROUP_ID = 3;
  IPMC_ENTRY_ATTR_COUNTER_ID = 4;
}

message CreateIpmcEntryRequest {
  option (sai_type) = OBJECT_TYPE_IPMC_ENTRY;
  IpmcEntry entry = 1;
  optional PacketAction packet_action = 2 [(attr_enum_value) = 1];
  optional uint64 output_group_id = 3 [(attr_enum_value) = 2];
  optional uint64 rpf_group_id = 4 [(attr_enum_value) = 3];
  optional uint64 counter_id = 5 [(attr_enum_value) = 4];
}

message CreateIpmcEntryResponse {}

message RemoveIpmcEntryRequest {
  IpmcEntry entry = 1;
}

message RemoveIpmcEntryResponse {}

message SetIpmcEntryAttributeRequest {
  IpmcEntry entry = 1;
  optional PacketAction packet_action = 2 [(attr_enum_value) = 1];
  optional uint64 output_group_id = 3 [(attr_enum_value) = 2];
  optional uint64 rpf_group_id = 4 [(attr_enum_value) = 3];
  optional uint64 counter_id = 5 [(attr_enum_value) = 4];
}

message SetIpmcEntryAttributeResponse {}

message GetIpmcEntryAttributeRequest {
  IpmcEntry entry = 1;
  repeated IpmcEntryAttr attr_type = 2;
}

message GetIpmcEntryAttributeResponse {
  IpmcEntryAttribute attr = 1;
}

service Ipmc {
  rpc CreateIpmcEntry(CreateIpmcEntryRequest)
      returns (CreateIpmcEntryResponse) {}
  rpc RemoveIpmcEntry(RemoveIpmcEntryRequest)
      returns (RemoveIpmcEntryResponse) {}
  rpc SetIpmcEntryAttribute(SetIpmcEntryAttributeRequest)
      returns (SetIpmcEntryAttributeResponse) {}
  rpc GetIpmcEntryAttribute(GetIpmcEntryAttributeRequest)
      returns (GetIpmcEntryAttributeResponse) {}
}
